src/domain/usecases/enrollments/CloseEnrollmentsUseCase.ts

```diff
@@ -0,0 +1,59 @@
+import _ from "lodash";
+import log from "utils/log";
+
+import { Async } from "domain/entities/Async";
+
+import { EventsRepository, EventsRepositoryParams } from "domain/repositories/enrollments/EventsRepository";
+import {
+    EnrollmentsRepository,
+    EnrollmentsRepositoryParams,
+    TrackerResponse,
+} from "domain/repositories/enrollments/EnrollmentsRepository";
+
+export class CloseEnrollmentsUseCase {
+    constructor(
+        private eventsRepository: EventsRepository,
+        private enrollmentsRepository: EnrollmentsRepository
+    ) {}
+
+    async execute(params: EventsRepositoryParams): Async<void> {
+        const events = await this.eventsRepository.getAll(params);
+
+        const enrollmentsIDs = _.uniqBy(events, "enrollment").flatMap(event => event.enrollment);
+
+        if (enrollmentsIDs.length === 0) {
+            log.info("No Enrollments for the provided date");
+            return;
+        }
+
+        const enrollmentsParams: EnrollmentsRepositoryParams = {
+            programId: params.programId,
+            orgUnitId: params.orgUnitId,
+        };
+
+        const enrollments = await this.enrollmentsRepository.getAll(enrollmentsParams);
+
+        const activeEnrollmentsIDs = enrollments
+            .filter(enrollment => enrollment.status === "ACTIVE")
+            .filter(enrollment => enrollmentsIDs.includes(enrollment.enrollment))
+            .map(enrollment => enrollment.enrollment);
+
+        if (activeEnrollmentsIDs.length === 0) {
+            log.info("No active Enrollments found for the provided date");
+            return;
+        }
+
+        log.info(`Found ${activeEnrollmentsIDs.length} active enrollments to close`);
+
+        activeEnrollmentsIDs.forEach(async enrollmentId => {
+            const report: TrackerResponse = await this.enrollmentsRepository.closeEnrollment(enrollmentId);
```

Comment:
That's a typical problem, forEach+await does not work as expected. ChatGPT gives a nice explanation as for why. We have the abstraction promiseMap for this scenario.

If we can finally post the enrollments in a single payload we won't need this, though.

Reply:
I remember now that this was touched upon in a dojo (i think). Anyway, we could use a promiseMap in 2.37 for performance reasons, [see](https://github.com/EyeSeeTea/d2-tools/pull/64#discussion_r1850949673). Maybe control the enrollment update method via an option.